### PR TITLE
Rollback2

### PR DESCRIPTION
**Pull request by jlmucb**
OK.  The tests work.  The first ReadNv fails and I have to figure out why.  There's probably about 50 more lines of code to write in tpm2_tao.go and tpm2.go before final checkin but it's safe to start.

### CODE REVIEW

File: go/tpm2/tpm2.go [hunk 1]
```diff
@@ -2427,105 +2428,292 @@ func ClientDecodeServerResponse(rw io.ReadWriter, protectorHandle Handle,
...
-	return nil, nil
+	cmdHdr, err := MakeCommandHeader(tagSESSIONS, 0, cmdUndefineSpace)
+	if err != nil {
+		return nil, errors.New("ConstructUndefineSpace failed")
+	}
+	auth := CreatePasswordAuthArea("", Handle(OrdTPM_RS_PW))
+	zero := uint16(0)
+	num_bytes := []interface{}{uint32(owner), uint32(handle), zero}
+	out, err := pack(num_bytes)
+	if err != nil {
```
Comment (sidtelang) [inline go/tpm2/tpm2.go]: return nil, err?

File: go/tao/tpm2_tao.go [hunk 2]
```diff
@@ -648,23 +656,92 @@ func (tt *TPM2Tao) Unseal(sealed []byte) (data []byte, policy string, err error)
...
+	_ = s.InitCounter(label, int64(0))
+	c, err := tpm2.GetCounter(s.rw, s.nvHandle, s.authString)
+	if err != nil {
+		return nil, err
+	}
+	err = tpm2.IncrementNv(s.rw, s.nvHandle, s.authString)
+	if err != nil {
+		return nil, err
+	}
+	if c % 2 == 0 {
```
Comment (sidtelang) [inline go/tao/tpm2_tao.go]: Not sure if I'm missing something but it appears to be this check should be c % 2 == 1
Say the nvram counter was say 2 before calling this function. As it is, it jumps to 5 (recall the init counter bumps it up by one too) as a result of the call which is not what we want.

File: go/tao/tpm2_tao.go [hunk 3]
```diff
@@ -648,23 +656,92 @@ func (tt *TPM2Tao) Unseal(sealed []byte) (data []byte, policy string, err error)
...
+	fmt.Printf("TPM2Tao.GetCounter\n")
+	_ = s.InitCounter(label, int64(0))
+	return tpm2.GetCounter(s.rw, s.nvHandle, s.authString)
 }
 
+// Note:  Tpm2 counters work differently from other counters.  On startup, you
+// can't read a counter value before you initialize it which you do by incrementing
+// it.  What we do is use the (countvalue+1)/2 as the counter in RollbackSeal and Unseal.
+// When you seal, // if the current counter is odd, you bump it twice and use the 
+// value (countvalue+1)/2 in the counter slot.  If the counter is even, you bump by 1.
```
Comment (sidtelang) [inline go/tao/tpm2_tao.go]: We're relying on the fact that the linux_host never calls InitCounter on its host tao directly?

File: go/tao/tpm2_tao.go [hunk 4]
```diff
@@ -648,23 +656,92 @@ func (tt *TPM2Tao) Unseal(sealed []byte) (data []byte, policy string, err error)
...
+	s.authString = "01020304"
+
+	return tpm2.InitCounter(s.rw, s.nvHandle, s.authString)
 }
 
 func (s *TPM2Tao) GetCounter(label string) (int64, error) {
-	fmt.Printf("tpm2.GetCounter\n") // REMOVE
-	return int64(0), errors.New("GetCounter for tpm2 not implemented")
+	fmt.Printf("TPM2Tao.GetCounter\n")
+	_ = s.InitCounter(label, int64(0))
```
Comment (jlmucb) [inline go/tao/tpm2_tao.go]: fixed

File: go/tao/tpm2_tao.go (hunk 3 above)
Comment (jlmucb) [inline go/tao/tpm2_tao.go]: We're doing this so we don't have to rely on InitCounter being called before GetCounter, or RollbackSeal/Unseal

File: go/tao/tpm2_tao.go (hunk 2 above)
Comment (jlmucb) [inline go/tao/tpm2_tao.go]: good catch

File: go/tpm2/tpm2.go (hunk 1 above)
Comment (jlmucb) [inline go/tpm2/tpm2.go]: fixed

**Pull request merged**